memgpt/data_types.py

```diff
@@ -0,0 +1,124 @@
+""" This module contains the data types used by MemGPT. Each data type must include a function to create a DB model. """
+import uuid
+from abc import abstractmethod
+from typing import Optional
+import numpy as np
+
+
+# Defining schema objects:
+# Note: user/agent can borrow from MemGPTConfig/AgentConfig classes
+
+
+class Record:
+    """
+    Base class for an agent's memory unit. Each memory unit is represented in the database as a single row.
+    Memory units are searched over by functions defined in the memory classes
+    """
+
+    def __init__(self, id: Optional[str] = None):
+        if id is None:
+            self.id = uuid.uuid4()
+        else:
+            self.id = id
+
+        assert isinstance(self.id, uuid.UUID), f"UUID {self.id} must be a UUID type"
+
+
+class Message(Record):
+    """Representation of a message sent from the agent -> user. Also includes function calls."""
+
+    def __init__(
+        self,
+        user_id: str,
+        agent_id: str,
+        role: str,
+        text: str,
+        model: str,  # model used to make function call
+        user: Optional[str] = None,  # optional participant name
+        created_at: Optional[str] = None,
+        function_name: Optional[str] = None,  # name of function called
```

Comment:
I think there's an important question here if we want to migrate our internal representations of messages to be more aligned with the new `tool` style instead of `function` style (official deprecated by OpenAI).

Functionally, I think the main difference to consider is that a message record/object can have multiple function/tool calls associated with it (instead of just one).

Additionally, tools get tagged with unique IDs when they are "generated" (not run, just params output by the LLM).

Then, in the "response" (role == `function/tool`), this unique ID is tagged to associate functions with their generated outputs. Previously, this association was done in a less structured way (basically, you assume that a message with role `function` is preceded by a message with role `assistant` that has `function != None`.

Previously I didn't think migrating to this internal representation was very important because we don't deal with parallel tool calling, but I think we might as well migrate now since we're making a big breaking change.

Reply:
This is what a tool response looks like:
<img width="551" alt="image" src="https://github.com/cpacker/MemGPT/assets/5475622/08f1f6f1-9728-46c0-bd1f-c0ac02a355f1">

vs a function response (no ID!):
<img width="391" alt="image" src="https://github.com/cpacker/MemGPT/assets/5475622/576c1f27-06b5-4a75-9c86-9507bbc5ccc5">

Reply:
Another reason to migrate to the tool style is that AFAIK the Assistants API only uses `tools`, no `function` deprecated backcompat.

Reply:
See #783